### ISSUE TITLE
[no-non-null-assertion] False positive with value that can have the value null

### ISSUE THREAD

**Issue by mdouglass** [typescript-eslint/typescript-eslint#1195]
**Repro**

```JSON
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": { "project": "./tsconfig.json", "sourceType": "module" },
  "plugins": ["@typescript-eslint"],
  "rules": {
    "@typescript-eslint/no-non-null-assertion": "error"
  }
}

```

```TS
function testFunction(_param: string | undefined): void { /* noop */ }
const value = 'test' as string | null | undefined
testFunction(value!)
```

**Expected Result**
No errors

**Actual Result**
3:14  error  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion

**Additional Info**
If the `!` is removed, then you get a compiler error:
Argument of type 'string | null | undefined' is not assignable to parameter of type 'string | undefined'.
  Type 'null' is not assignable to type 'string | undefined'.ts(2345)

**Versions**

| package | version |
| ---------------------------------- | ------- |
| @typescript-eslint/eslint-plugin | 2.6.1 |
| @typescript-eslint/parser | 2.6.1 |
| TypeScript | 3.7.2 |
| ESLint | 6.6.0 |
| node | 13.1.0 |
| npm | 6.12.1 |

**Comment by bradzacher** [comment]
You misunderstand the intention of the rule - it's not intended to be a check for "is it a valid usage"

It's an outright ban of the operator, because it is considered to be unsafe.

I.e. in your code, you could have

```ts
function testFunction(param: string | undefined): number {
  if (param !== undefined) { // strict check because typings enforce it is string | undefined onyl
    return param.length
  }
  return -1;
}

const value = null as string | null | undefined
testFunction(value!)
// ERROR - cannot read length of null
```

If you want a rule that checks for unnecessary non-null operators, then you want `no-unnecessary-type-assertion`

**Comment by mdouglass** [comment]
Damnit, I knew that and am just so heads down in this tslint -> eslint migration that I spaced.  Sorry to bug you with it. :)
